Point at expressions where inference refines an unexpected type

estebank opened the pull request:
Fix #106355. Fix #14007. (!)

```
error[E0308]: mismatched types
  --> src/test/ui/type/type-check/point-at-inference.rs:12:9
   |
9  |         foo.push(i);
   |                  - this is of type `&{integer}`, which makes `foo` to be inferred as `Vec<&{integer}>`
...
12 |     bar(foo);
   |     --- ^^^ expected `i32`, found `&{integer}`
   |     |
   |     arguments to this function are incorrect
   |
   = note: expected struct `Vec<i32>`
              found struct `Vec<&{integer}>`
note: function defined here
  --> src/test/ui/type/type-check/point-at-inference.rs:2:4
   |
2  | fn bar(_: Vec<i32>) {}
   |    ^^^ -----------
help: consider dereferencing the borrow
   |
9  |         foo.push(*i);
   |                  +
```

estebank committed:
Point at expressions where inference refines an unexpected type

Address #106355.

compiler-errors commented on compiler/rustc_hir_typeck/src/demand.rs:
```diff
@@ -205,6 +210,157 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
+                    return t;
+                }
+                match *t.kind() {
+                    ty::Infer(ty::TyVar(_) | ty::FreshTy(_)) => {
```
We should never have `FreshTy`, etc types show up in typeck itself. Do you need this due to a UI test failing or something?

compiler-errors commented on compiler/rustc_hir_typeck/src/demand.rs:
```diff
@@ -53,7 +57,8 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
             || self.suggest_into(err, expr, expr_ty, expected)
-            || self.suggest_floating_point_literal(err, expr, expected);
+            || self.suggest_floating_point_literal(err, expr, expected)
+            || self.point_inference_types(err, expr);
```
I'd prefer if we do something like:

```rust
let suggestion_applied = ...;

// If no suggestion is applied, point at where the inference happened.
if !suggestion_applied {
  self.point_at_inference_types(...);
}
```

That is, it feels somewhat out of place to chain this non-suggestion method with a bunch of other suggestion methods here.

compiler-errors commented on compiler/rustc_hir_typeck/src/demand.rs:
```diff
@@ -205,6 +210,157 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
+            ..
+        })) = map.find(parent) else { return false; };
+
+        let ty = self.node_ty(init.hir_id);
```
`node_ty_opt` is a better choice here, just in case.

compiler-errors commented on src/test/ui/rfc-2497-if-let-chains/ensure-that-let-else-does-not-interact-with-let-chains.stderr:
```diff
@@ -120,6 +126,12 @@ LL |     let Some(n) = opt && n == 1 else {
 error[E0308]: mismatched types
   --> $DIR/ensure-that-let-else-does-not-interact-with-let-chains.rs:15:19
    |
+LL |     let opt = Some(1i32);
+   |               ---------- here the type of `opt` is inferred to be `Option<_>`
+LL |
+LL |     let Some(n) = opt else {
+   |                   --- here the type of `opt` is inferred to be `Option<i32>`
```
Hm, this doesn't look correct :thinking:

compiler-errors replied:
To be more specific, the problem here is that we record the return type of the `Some` constructor before we unify the arguments.

Try adding a `resolve_vars_if_possible` to the type we pass to `write_ty` in [`check_expr_with_expectation_and_args`](https://doc.rust-lang.org/nightly/nightly-rustc/src/rustc_hir_typeck/expr.rs.html#264). That might help!

compiler-errors commented on src/test/ui/span/coerce-suggestions.stderr:
```diff
@@ -63,7 +63,10 @@ error[E0308]: mismatched types
+   |         ^^^^^^^^^^^^^^
+   |         |
+   |         expected `&mut String`, found struct `String`
+   |         here the type of `res` is inferred to be `String`
```
Hm, if the "here the type of .." span overlaps the blame span of the type error, we probably shouldn't emit it.

Also `res`? Where's that coming from -- within the macro? We probably shouldn't mention macro-local variables.

compiler-errors commented on src/test/ui/suggestions/issue-86100-tuple-paren-comma.stderr:
```diff
@@ -49,6 +49,8 @@ LL |     let _s = S { _s: ("abc".to_string(),) };
 error[E0308]: mismatched types
   --> $DIR/issue-86100-tuple-paren-comma.rs:23:22
    |
+LL |     let t = (1, 2);
```
I'm not exactly sure how to express this, but I'd prefer if we didn't emit this for literal expressions or expression that are almost "literal-like" -- this suggestion doesn't seem to add much value, but I'm not exactly sure if I can exactly figure why.

estebank replied:
I do think that there's some value to *show* the code where the initialization took place, so that the user doesn't have to go immediately back to their editor to understand what happened. Having said that, I'm ok eliding some of these if you feel that's a better idea.

compiler-errors replied:
Hm, I just feel like this PR is stated to do something different (showing where an inferece type is *refined* to a specific value) but then also increases verbosity in a separate situation (pointing out where a local is initialized) as a side-effect.

I think if we *do* want to point out where a local is initialized in more cases, that should be done separately, so we can properly gauge its impact and maybe have a deeper discussion about it.

compiler-errors commented on compiler/rustc_hir_typeck/src/demand.rs:
```diff
@@ -205,6 +210,157 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
+                    } else {
+                        self.node_ty(rcvr.hir_id)
+                    };
+                    let ty = eraser.fold_ty(ty);
+                    if ty.references_error() {
+                        break;
+                    }
+                    if ty != prev {
```
This equality is suspicious -- can you explain what this erasing is for in detail?

estebank replied:
This is what I was seeing:

```
fn bar(_: Vec<i32>) {}
fn main() {
    let v: Vec<i32> = vec![1, 2, 3, 4, 5];
    let mut foo = vec![];
//                ^^^^^^ node_ty Vec<_1>
    baz(&foo);
//       ^^^ node_ty Vec<_2>
    for i in &v {
        foo.push(i);
//      ^^^ node_ty Vec<_3> 
    }
    bar(foo);
//      ^^^ node_ty:Vec<&i32>
}
```
Because these all had different ty::Infer, they weren't equal, but I couldn't use can_eq instead because that would have coerced all to `Vec<&i32>` and I wouldn't be able to show any of these spans, only the first one. I needed a mechanism to look at the node_ty changing from expression to expression only when the type was "narrowed" to a more specific one.

compiler-errors replied:
Hm, I guess I understand you mean. Thanks for the example.

Ideally, we'd fold under something a bit better like [`InferCtxt::root_var`](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_infer/infer/struct.InferCtxt.html#method.root_var) to turn `_1`, `_2` and `_3` into a common infer variable.

compiler-errors commented on compiler/rustc_hir_typeck/src/demand.rs:
```diff
@@ -205,6 +210,157 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
+                {
+                    let ty = if let Ok(m) = self.lookup_method(ty, s, span, expr, rcvr, args) {
+                        // We get the self type from `lookup_method` because the `rcvr` node
+                        // type will not have had any adjustments from the fn arguments.
```
What about `node_ty_adjusted`? Doesn't that show us the adjusted type of the rcvr expr?

If that doesn't work, then I'd really prefer if you avoided this extra lookup with something by using the already computed method def-id for a hir expr by using a combination of these two methods:
1. https://doc.rust-lang.org/nightly/nightly-rustc/rustc_middle/ty/typeck_results/struct.TypeckResults.html#method.type_dependent_def
2. https://doc.rust-lang.org/nightly/nightly-rustc/rustc_middle/ty/typeck_results/struct.TypeckResults.html#method.node_substs

estebank replied:
`expr_ty_adjusted` gives us the following:

```rust
% rustc +dev f57.rs -Zverbose
error[E0308]: mismatched types
  --> f57.rs:12:9
   |
6  |     let mut foo = vec![];
   |                   ------ here the type of `foo` is inferred to be `Vec<_#16t>`
...
9  |         foo.push(i);
   |         --- here the type of `foo` is inferred to be `&ReErased mut Vec<_#0t>`
10 |     }
11 |     baz(&foo);
   |          --- here the type of `foo` is inferred to be `Vec<&ReErased _#0i>`
12 |     bar(foo);
   |     --- ^^^ expected `i32`, found `&_#0i`
   |     |
   |     arguments to this function are incorrect
   |
   = note: expected struct `Vec<i32>`
              found struct `Vec<&'_#14r _#0i>`
```

Notice that `_#0t` is the inference variable on `foo`, so `i` hasn't affected it yet. I also still need to remove the self receiver borrow.

estebank replied:
Interestingly, with the following:

```rust
                    && let Some(def_id) = self.typeck_results.borrow().type_dependent_def_id(expr.hir_id)
                    && let Some(substs) = self.typeck_results.borrow().node_substs_opt(expr.hir_id)
                {
                    let sig = self.tcx.fn_sig(def_id);
                    let method_ty = tcx.mk_fn_def(def_id, substs);
                    err.note(&format!("\n{:?}\n{:?}", sig, method_ty.kind()));
```

the output is

```
   = note:
              Binder(([&ReLateBound(DebruijnIndex(0), BoundRegion { var: 0, kind: BrNamed(DefId(5:8965 ~ alloc[b102]::vec::{impl#1}::push::'_), '_) }) mut std::vec::Vec<T, A>, T]; c_variadic: false)->(), [Region(BrNamed(DefId(5:8965 ~ alloc[b102]::vec::{impl#1}::push::'_), '_))])
              FnDef(DefId(5:6488 ~ alloc[b102]::vec::{impl#1}::push), [_#37t, _#38t])
```

which looks to me to have only inference variables, because the `node_substs` for `expr.hir_id` do not have the changes introduced by `push(i)` yet.

I also wanted to reuse as much of the existing machinery as possible, as I don't want it to diverge after a refactor. I'm looking if calling `prove_op` directly is feasible from here instead, just to see if the type is in `pick` already.

estebank replied:
Do let me know what you think of the current approach. It's not quite what you asked, but I think it is robust.

compiler-errors commented on compiler/rustc_hir_typeck/src/demand.rs:
```diff
@@ -205,6 +210,157 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
+        let hir::ExprKind::Path(hir::QPath::Resolved(None, p)) = expr.kind else { return false; };
+        let [hir::PathSegment { ident, args: None, .. }] = p.segments else { return false; };
+        let hir::def::Res::Local(hir_id) = p.res else { return false; };
+        let Some(node) = map.find(hir_id) else { return false; };
```
nit: I think the usages of `find` are a bit defensive here -- I think you can just use `get` here.

Or at least this can be combined with the expression to make this shorter and easier to understand.

compiler-errors commented on compiler/rustc_hir_typeck/src/demand.rs:
```diff
@@ -205,6 +210,157 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
+                    }
+                    prev = ty;
+                }
+                if ex.hir_id == expr.hir_id {
```
nit: `ex` and `expr` is pretty confusing. can you give one of these a more descriptive name?

estebank replied:
Yeah, I need to do another pass for naming

compiler-errors commented on compiler/rustc_hir_typeck/src/demand.rs:
```diff
@@ -205,6 +210,157 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
+                        // We get the self type from `lookup_method` because the `rcvr` node
+                        // type will not have had any adjustments from the fn arguments.
+                        let ty = m.sig.inputs_and_output[0];
+                        match ty.kind() {
+                            // Remove one layer of references to account for `&mut self` and
+                            // `&self`, so that we can compare it against the binding.
+                            ty::Ref(_, ty, _) => *ty,
+                            _ => ty,
+                        }
```
Hm, I suspect this will mess up cases where the rcvr type is `&T` but it's not being autoref'd...

estebank replied:
Yeah, I was considering peeking at the fn def to do the right thing always, but hopefully node_ty_adjusted does what I want it to (but I'm not holding my breath).

compiler-errors commented on src/test/ui/closures/closure-return-type-mismatch.stderr:
```diff
@@ -1,6 +1,8 @@
 error[E0308]: mismatched types
   --> $DIR/closure-return-type-mismatch.rs:7:9
    |
+LL |         let a = true;
```
Hm, I'm not convinced that pointing out the type of a literal at the place a variable is initialized is very useful compared to other expressions or places further away from where the variable is declared...

estebank replied:
I have two options: do not show any span labels if the only thing it's doing is pointing at the declaration, or ignore literal expressions. Either way, it might not be interesting as a way to understand the inference steps, *but* it works as a "found `bool` because of this init value".

compiler-errors commented on compiler/rustc_hir_typeck/src/demand.rs:
```diff
@@ -205,6 +210,157 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
+        let map = self.tcx.hir();
+
+        // Hack to make equality checks on types with inference variables and regions useful.
+        struct TypeEraser<'tcx> {
+            tcx: TyCtxt<'tcx>,
+        }
```
This can use a [`BottomUpFolder`](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_middle/ty/fold/struct.BottomUpFolder.html) probably

estebank replied:
Nice! I hadn't seen that when it was added, I'm guessing to support TAIT?

estebank commented on compiler/rustc_hir_typeck/src/demand.rs:
```diff
@@ -205,6 +210,157 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
+        let tcx = self.tcx;
+        let map = self.tcx.hir();
+
+        // Hack to make equality checks on types with inference variables and regions useful.
```
This is so that I can `ty == ty`. It's like if `can_eq` *didn't* coerce `_` and `{integer}`, so I can compare `Ty<_r#1>` with `Ty<_r#2>` as equal, but `Ty<_>` with `Ty<X>` are unequal. In effect this is the same as doing ty.to_string() == ty.to_string() (except when `-Zverbose` is set).

compiler-errors replied:
Is this *just* for region variables? I'm not sure if I understand why we would want to treat *type* (or const) inference variables as equal?

estebank replied:
No, for region variables I can just `erase_regions`, this is for type inference variables. I want to be able to say "are these two types the same *including* inference variables?" As in `Ty<_> == Ty<_>` (regardless of what `_` is, they can have different `TyVid`), but also `Ty<_> != Ty<()>`.

estebank committed:
Skip macros to avoid talking about bindings the user can't see

estebank committed:
More eagerly resolve expr `ty`s before writing them

This allows the expressions to have more accurate types
when showing inference steps.

compiler-errors commented on compiler/rustc_hir_typeck/src/expr.rs:
```diff
@@ -234,6 +234,7 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
             ) => self.check_expr_path(qpath, expr, args),
             _ => self.check_expr_kind(expr, expected),
         });
+        let ty = self.resolve_vars_if_possible(ty);
```
super tiny nit: Not sure why this commit title mentions "let chain `ty`"s -- this resolves all checked expr types :laughing:

estebank commented on compiler/rustc_hir_typeck/src/demand.rs:
```diff
@@ -205,6 +210,157 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
+        let hir::Node::Pat(pat) = node else { return false; };
+        let parent = map.get_parent_node(pat.hir_id);
+        let Some(hir::Node::Local(hir::Local {
+            ty: None,
```
I just realized that this is overly restrictive, as you can write `let x: Vec<_> = vec![]`.

estebank replied:
This point is now moot as we no longer point at all the spans, just the expression where the found type was fully formed.

estebank committed:
review comments: do not always point at init expr

estebank committed:
Tweak output

- Only point at a the single expression where the found type was first
  inferred.
- Find method call argument that might have caused the found type to be
  inferred.
- Provide structured suggestion.
- Apply some review comments.
- Tweak wording.

estebank committed:
Formatting

estebank committed:
Use `BottomUpFolder`

estebank commented:
@compiler-errors I think I addressed all of your previous comments

estebank committed:
Account for type error on method arg caused by earlier inference

```rust
fn main() {
    let v = Vec::new();
    v.push(0);
    v.push(0);
    v.push("");
}
```

now produces

```
error[E0308]: mismatched types
  --> $DIR/point-at-inference-3.rs:6:12
   |
LL |     v.push(0);
   |            - this is of type `{integer}`, which makes `v` to be inferred as `Vec<{integer}>`
...
LL |     v.push("");
   |       ---- ^^ expected integer, found `&str`
   |       |
   |       arguments to this function are incorrect
   |
note: associated function defined here
  --> $SRC_DIR/alloc/src/vec/mod.rs:LL:COL
```

compiler-errors commented on compiler/rustc_hir_typeck/src/demand.rs:
```diff
@@ -205,6 +213,217 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
+                }
+                c.errors = vec![];
+                let _ = c.relate(def_self_ty, expected);
+                for error in c.errors {
+                    if let TypeError::Sorts(error) = error {
+                        param_expected.insert(error.expected, error.found);
+                    }
+                }
```
I'm really confused what's going on here. Can you explain?

compiler-errors replied:
Are you using this to collect a mapping from parameter type to substitution? If so, I'm not sure I'm a fan of using `CollectAllMismatches` for that purpose...

estebank replied:
That's what I'm doing, trying to collect `(T, `i32`, `&i32`). If you can think of a better approach, I'm all ears.

compiler-errors commented on compiler/rustc_hir_typeck/src/demand.rs:
```diff
@@ -205,6 +213,217 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
+                    let Some(expected) = param_expected.get(param) else { continue; };
+                    let Some(found) = param_found.get(param) else { continue; };
+                    if self.can_eq(self.param_env, *arg_ty, *found).is_err() { continue; }
+                    self.suggest_deref_ref_or_into(err, arg, *expected, *found, None);
```
Why not suggest all type mismatch suggestions?

estebank replied:
No good reason

estebank committed:
Suggest changing argument on type error

estebank committed:
review comment: potentially produce more suggestions for arg type mismatch

compiler-errors commented on compiler/rustc_hir_typeck/src/demand.rs:
```diff
@@ -205,6 +213,217 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
+                            err.span_label(
+                                arg.span,
+                                &format!(
+                                    "this is of type `{arg_ty}`, which makes `{ident}` to be \
```
```suggestion
                                    "this is of type `{arg_ty}`, which causes `{ident}` to be \
```

compiler-errors reviewed (commented):
one nit: "make" here is a synonym of "cause" but I'm not sure if it works with the same construction?

r=me -- I may take a stab at this method code a bit later, but I don't want to block you and I'm still trying to think of a way to avoid the`CollectAllMismatches` usage here.

estebank commented on compiler/rustc_hir_typeck/src/demand.rs:
```diff
@@ -205,6 +213,217 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
+            {
+                // We special case methods, because they can influence inference through the
+                // call's arguments and we can provide a more explicit span.
+                let sig = self.tcx.fn_sig(def_id);
```
I'm using this to get the type without any substs. I'm using this as the "bare" type against which I can look for type parameters, for example. All I'm using it is for `def_self_ty`, which is meant to be `Vec<T, A>`, not `Vec<_1>`. Given that, what is the use of using `bound_fn_sig(...).subst(..)` for here?

compiler-errors replied:
Chronologically, I made this comment before https://github.com/rust-lang/rust/pull/106400#discussion_r1061997310

estebank committed:
review comments: reword

estebank commented:
@bors r=compiler-errors

bors commented:
:pushpin: Commit 25cf71e1be5a2166c11334d91f6f047501ed3264 has been approved by `compiler-errors`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=compiler-errors 25cf71e1be5a2166c11334d91f6f047501ed3264 -->
<!-- homu: {"type":"Approved","sha":"25cf71e1be5a2166c11334d91f6f047501ed3264","approver":"compiler-errors","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

bors commented:
:hourglass: Testing commit 25cf71e1be5a2166c11334d91f6f047501ed3264 with merge cfbc12774406f0ed9a269e2c28275de9e0a45330...
<!-- homu: {"type":"BuildStarted","head_sha":"25cf71e1be5a2166c11334d91f6f047501ed3264","merge_sha":"cfbc12774406f0ed9a269e2c28275de9e0a45330"} -->

bors commented:
:broken_heart: Test failed - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/3845584220/jobs/6549967971)
<!-- homu: {"type":"BuildFailed","builder_url":"https://github.com/rust-lang-ci/rust/actions/runs/3845584220/jobs/6549967971","builder_name":"checks-actions"} -->

rust-log-analyzer commented:
The job **`mingw-check`** failed! Check out the build log: [(web)](https://github.com/rust-lang-ci/rust/actions/runs/3845584220/jobs/6549901445) [(plain)](https://github.com/rust-lang-ci/rust/commit/cfbc12774406f0ed9a269e2c28275de9e0a45330/checks/10455630069/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
[RUSTC-TIMING] rustc_borrowck test:false 4.283
error[E0599]: no method named `get_parent_node` found for struct `rustc_middle::hir::map::Map` in the current scope
   --> compiler/rustc_hir_typeck/src/demand.rs:229:26
    |
229 |         let parent = map.get_parent_node(pat.hir_id);

error[E0599]: no method named `get_parent_node` found for struct `rustc_middle::hir::map::Map` in the current scope
   --> compiler/rustc_hir_typeck/src/demand.rs:288:30
    |
    |
288 |             let parent = map.get_parent_node(binding.hir_id);

[RUSTC-TIMING] rustc_borrowck test:true 4.495
[RUSTC-TIMING] rustc_codegen_llvm test:true 2.571
[RUSTC-TIMING] rustc_codegen_llvm test:false 2.653
```

</details>

estebank commented:
I'm very confused by this merge conflict, pulling and rebasing still builds 👀

estebank commented:
```
$ git log HEAD^^
commit 14298997005486529e7a97b2ab9e9e376d6fdc72
Merge: 03b9e1d154e e048ee2ac1c
Author: bors <bors@rust-lang.org>
Date:   Thu Jan 5 06:47:02 2023 +0000

    Auto merge of #106482 - compiler-errors:rollup-g7n1p39, r=compiler-errors
    
    Rollup of 6 pull requests
    
    Successful merges:
    
     - #105846 (Account for return-position `impl Trait` in trait in `opt_suggest_box_span`)
     - #106385 (Split `-Zchalk` flag into `-Ztrait-solver=(classic|chalk|next)` flag)
     - #106403 (Rename `hir::Map::{get_,find_}parent_node` to `hir::Map::{,opt_}parent_id`, and add `hir::Map::{get,find}_parent`)
     - #106462 (rustdoc: remove unnecessary wrapper around sidebar and mobile logos)
     - #106464 (Update Fuchsia walkthrough with new configs)
     - #106478 (Tweak wording of fn call with wrong number of args)
    
    Failed merges:
    
    r? `@ghost`
    `@rustbot` modify labels: rollup
```

🤷

estebank committed:
fix rebase

estebank commented:
@bors r=compiler-errors

bors commented:
:pushpin: Commit f98f2fcb5e4985bd7b12bae4ec03f2d96197fb2b has been approved by `compiler-errors`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=compiler-errors f98f2fcb5e4985bd7b12bae4ec03f2d96197fb2b -->
<!-- homu: {"type":"Approved","sha":"f98f2fcb5e4985bd7b12bae4ec03f2d96197fb2b","approver":"compiler-errors","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

bors commented:
:hourglass: Testing commit f98f2fcb5e4985bd7b12bae4ec03f2d96197fb2b with merge 6660916f5ebc99a31603558493a43b72f7189242...
<!-- homu: {"type":"BuildStarted","head_sha":"f98f2fcb5e4985bd7b12bae4ec03f2d96197fb2b","merge_sha":"6660916f5ebc99a31603558493a43b72f7189242"} -->

bors commented:
:boom: Test timed out
<!-- homu: {"type":"TimedOut"} -->

estebank commented:
@bors retry

rust-log-analyzer commented:
A job failed! Check out the build log: [(web)](https://github.com/rust-lang-ci/rust/actions/runs/3849368948/jobs/6562134889) [(plain)](https://github.com/rust-lang-ci/rust/commit/6660916f5ebc99a31603558493a43b72f7189242/checks/10471857017/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain

```

</details>
